component/storage/leveldb/leveldb.go

```diff
@@ -0,0 +1,503 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package leveldb
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/syndtr/goleveldb/leveldb"
+
+	"github.com/hyperledger/aries-framework-go/spi/storage"
+)
+
+// TODO (#2605) Implement all methods in this file. Currently the only methods implemented are the ones directly
+//  needed by aries-framework-go. It's possible that someone extending aries may find the other methods useful.
+
+const (
+	pathPattern = "%s-%s"
+
+	tagMapKey                       = "TagMap"
+	storeConfigKey                  = "StoreConfig"
+	expressionTagNameOnlyLength     = 1
+	expressionTagNameAndValueLength = 2
+	invalidQueryExpressionFormat    = `"%s" is not in a valid expression format. ` +
+		"it must be in the following format: TagName:TagValue"
+)
+
+// Provider leveldb implementation of storage.Provider interface.
+type Provider struct {
+	dbPath string
+	dbs    map[string]*store
+	lock   sync.RWMutex
+}
+
+type closer func(storeName string)
+
+type tagMapping map[string]map[string]struct{} // map[TagName](Set of database Keys)
```

Comment:
Note: LevelDB has no built in indexing mechanism, so the tag/indexing stuff has to be handled in the provider/store implementation directly.

Reply:
I haven't looked closely at all this storage work, but does this imply then that each individual node in a cluster has a different view of `tagMapping`? If so then that's not good and sounds like it should be stored in LevelDB.